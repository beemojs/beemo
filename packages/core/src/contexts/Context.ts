import camelCase from 'lodash/camelCase';
import trim from 'lodash/trim';
import { PrimitiveType } from '@boost/args';
import { FilePath,Path } from '@boost/common';
import { Context as BaseContext } from '@boost/pipeline';
import { Arguments, Argv } from '../types';

export interface ConfigPath {
  driver: string;
  path: Path;
}

export default class Context<
  O extends object = {},
  P extends PrimitiveType[] = PrimitiveType[]
> extends BaseContext {
  // Parsed command line arguments as an object
  args: Arguments<O, P>;

  // Raw command line arguments as a list of strings
  argv: Argv;

  // List of driver configs currently registered
  configPaths: ConfigPath[] = [];

  // Current working directory
  cwd!: Path;

  // Absolute path to the configuration module
  configModuleRoot!: Path;

  // Absolute path to the folder containing `package.json` (Yarn workspaces) or `lerna.json`
  workspaceRoot!: Path;

  // List of paths (with trailing glob star) for each defined workspace
  workspaces: FilePath[] = [];

  constructor(args: Arguments<O, P>, argv: Argv = []) {
    super();

    this.args = args;
    this.argv = argv;

    // TODO MIGRATE
    // @ts-expect-error
    this.addArg = this.addParam.bind(this);
    // @ts-expect-error
    this.addArgs = this.addParams.bind(this);
  }

  /**
   * Add a config path for the defined driver.
   */
  addConfigPath(driverName: string, path: Path): this {
    this.configPaths.push({
      driver: driverName,
      path,
    });

    return this;
  }

  /**
   * Add an option argument to both the args object and argv list.
   * If the option is not supported, it will be added to the unknown options object.
   */
  addOption(arg: string, defaultValue: PrimitiveType = true, useEquals: boolean = false): this {
    let option = arg;
    let value = defaultValue;

    if (option.includes('=')) {
      [option, value] = option.split('=');
      useEquals = true;
    }

    let name = trim(option, '-');

    if (name.startsWith('no-')) {
      name = name.slice(3);
      value = false;
    }

    const optName = camelCase(name);

    if (optName in this.args.options) {
      // @ts-expect-error Allow this
      this.args.options[optName] = value;
    } else {
      this.args.unknown[optName] = String(value);
    }

    if (typeof value === 'boolean' || !value) {
      this.argv.push(option);
    } else if (useEquals) {
      this.argv.push(`${option}=${value}`);
    } else {
      this.argv.push(option, String(value));
    }

    return this;
  }

  /**
   * Add multiple boolean option arguments.
   */
  addOptions(args: string[]): this {
    args.forEach((arg) => {
      this.addOption(arg);
    });

    return this;
  }

  /**
   * Add a parameter to the argv list.
   */
  addParam(arg: string): this {
    this.args.params.push(arg);
    this.argv.push(arg);

    return this;
  }

  /**
   * Add multiple parameters.
   */
  addParams(args: string[]): this {
    args.forEach((arg) => {
      this.addParam(arg);
    });

    return this;
  }

  /**
   * Find a configuration path by file name.
   */
  findConfigByName(name: string): ConfigPath | undefined {
    return this.configPaths.find(
      (config) => String(config.path).endsWith(name) || config.driver === name,
    );
  }

  /**
   * Return a configured option value by name, or a fallback value if not found.
   */
  getOption<K extends keyof O>(name: K, fallback?: O[K]): O[K] | null {
    return this.args.options[name] || fallback || null;
  }

  /**
   * Return either a configured option value, or an unknown option value,
   * or null if not found.
   */
  getRiskyOption(name: string): PrimitiveType | null {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return (this.getOption(name as keyof O) ?? this.args.unknown[name] ?? null) as any;
  }
}
