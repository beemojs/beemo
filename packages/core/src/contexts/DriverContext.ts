import ConfigContext from './ConfigContext';
import Driver from '../Driver';
import { Arguments, Argv, StdioType } from '../types';

export interface DriverContextOptions {
  concurrency: number;
  graph: boolean;
  stdio: StdioType;
  workspaces: string;
}

export default class DriverContext extends ConfigContext<DriverContextOptions> {
  // Name defined on the plugin (kebab case)
  driverName: string = '';

  // Nested list of argv for each parallel execution
  parallelArgv: Argv[] = [];

  // The primary driver that initiated the pipeline
  primaryDriver: Driver;

  constructor(args: Arguments<DriverContextOptions>, driver: Driver, parallelArgv: Argv[] = []) {
    super(args);

    this.driverName = driver.name;
    this.parallelArgv = parallelArgv;
    this.primaryDriver = driver;

    // Add primary driver to driver list
    this.drivers.add(driver);
  }

  /**
   * Add a parallel command with additional argv.
   */
  addParallelCommand(argv: Argv): this {
    this.parallelArgv.push(argv);

    return this;
  }
}
