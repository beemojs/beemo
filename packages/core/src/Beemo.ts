import fs from 'fs-extra';
import path from 'path';
import { Event } from '@boost/event';
import { bool, number, string, shape } from 'optimal';
import CleanupConfigsRoutine from './routines/CleanupConfigsRoutine';
import ResolveConfigsRoutine from './routines/ResolveConfigsRoutine';
import RunDriverRoutine from './routines/RunDriverRoutine';
import Driver from './Driver';
import Context from './contexts/Context';
import DriverContext from './contexts/DriverContext';
import { Argv, Execution, BeemoConfig, BeemoPluginRegistry, DriverOptions } from './types';

export function configBlueprint() {
  return {
    configure: shape({
      cleanup: bool(false),
      parallel: bool(true),
    }),
    execute: shape({
      concurrency: number(),
      graph: bool(true),
    }),
    module: process.env.BEEMO_CONFIG_MODULE
      ? string(process.env.BEEMO_CONFIG_MODULE)
      : string().required(),
  };
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export default class Beemo<T = any> extends Tool<BeemoPluginRegistry, BeemoConfig<T>> {
  onRunDriver = new Event<[DriverContext, Driver<object, DriverOptions>]>('run-driver');

  constructor(argv: Argv, binName?: string, testingOnly: boolean = false) {
    super(
      {
        appName: 'beemo',
        appPath: path.join(__dirname, '..'),
        configBlueprint: configBlueprint(),
        configName: binName || 'beemo',
        scoped: true,
      },
      argv,
    );

    // Abort early for testing purposes
    if (testingOnly) {
      return;
    }

    this.initialize();

    // Set footer after messages have been loaded
    const footer = this.msg('app:poweredBy', { version });

    this.options.footer = `\n${this.isCI() ? '' : 'ðŸ¤–  '}${footer}`;
  }

  /**
   * Execute all routines for the chosen driver.
   */
  async runDriver(
    args: DriverContext['args'],
    driverName: string,
    parallelArgv: Argv[] = [],
  ): Promise<Execution[]> {
    const driver = this.getPlugin('driver', driverName);
    const context = this.prepareContext(new DriverContext(args, driver, parallelArgv));
    const version = driver.getVersion();

    this.onRunDriver.emit([context, driver], driverName);

    this.debug('Running with %s v%s driver', driverName, version);

    const pipeline = this.startPipeline(context)
      .pipe(new ResolveConfigsRoutine('config', this.msg('app:configGenerate')))
      .pipe(
        new RunDriverRoutine(
          'driver',
          this.msg('app:driverRun', {
            name: driver.metadata.title,
            version,
          }),
        ),
      );

    // Only add cleanup routine if we need it
    if (this.config.configure.cleanup) {
      pipeline.pipe(new CleanupConfigsRoutine('cleanup', this.msg('app:cleanup')));
    }

    return pipeline.run(driverName);
  }

  /**
   * Delete config files if a process fails.
   */
  private handleCleanupOnFailure(code: number, context: Context) {
    if (code === 0) {
      return;
    }

    // Must not be async!
    if (Array.isArray(context.configPaths)) {
      context.configPaths.forEach((config) => {
        fs.removeSync(config.path.path());
      });
    }
  }
}
