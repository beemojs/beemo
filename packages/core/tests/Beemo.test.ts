import path from 'path';
import fs from 'fs-extra';
import optimal from 'optimal';
import Beemo from '../src/Beemo';
import Context from '../src/contexts/Context';
import DriverContext from '../src/contexts/DriverContext';
// @ts-ignore
import bootstrapIndex from '../../../tests';

jest.mock(
  '@boost/core/lib/Pipeline',
  () =>
    function PipelineMock() {
      // @ts-ignore
      this.pipe = jest.fn(() => this);
      // @ts-ignore
      this.run = jest.fn(() => this);
    },
);

jest.mock('../../../tests', () => jest.fn());

const root = path.join(__dirname, '../../../tests');

describe('Beemo', () => {
  const oldExistsSync = fs.existsSync;
  const oldReadJsonSync = fs.readJsonSync;
  const oldRemoveSync = fs.removeSync;
  let beemo: Beemo;
  let onSpy: jest.Mock;

  beforeEach(() => {
    beemo = new Beemo(['foo', 'bar'], '', mockTool());
    beemo.moduleRoot = root;
    beemo.tool.options.root = root;

    // Stop `exit` event from firing
    onSpy = jest.fn();
    beemo.tool.on = onSpy;

    fs.existsSync = jest.fn();
    fs.readJsonSync = jest.fn();
    fs.removeSync = jest.fn();
  });

  afterEach(() => {
    fs.existsSync = oldExistsSync;
    fs.readJsonSync = oldReadJsonSync;
    fs.removeSync = oldRemoveSync;
  });

  it('sets argv', () => {
    expect(beemo.argv).toEqual(['foo', 'bar']);
  });

  describe('bootstrapConfigModule()', () => {
    beforeEach(() => {
      (bootstrapIndex as jest.Mock).mockReset();
      beemo.tool.config.module = '@local';
    });

    it('does nothing if no index file', () => {
      (fs.existsSync as jest.Mock).mockImplementation(() => false);

      beemo.bootstrapConfigModule();

      expect(bootstrapIndex).not.toHaveBeenCalled();
    });

    it('calls bootstrap with tool if index exists', () => {
      (fs.existsSync as jest.Mock).mockImplementation(() => true);

      beemo.bootstrapConfigModule();

      expect(bootstrapIndex).toHaveBeenCalledWith(beemo.tool);
    });
  });

  describe('createConfigFiles()', () => {
    beforeEach(() => {
      // @ts-ignore
      beemo.tool.getPlugin = (type, name) => createTestDriver(name);
      beemo.tool.getPlugins = () => [
        createTestDriver('foo'),
        createTestDriver('bar'),
        createTestDriver('baz'),
      ];
    });

    it('triggers `init-driver` event with context for the first driver (primary)', async () => {
      const spy = jest.spyOn(beemo.tool, 'emit');

      await beemo.createConfigFiles(MOCK_CONFIG_ARGS, ['foo']);

      expect(spy).toHaveBeenCalledWith('foo.init-driver', [
        expect.objectContaining({
          argv: ['foo', 'bar'],
          drivers: expect.anything(),
        }),
        expect.objectContaining({ name: 'foo' }),
      ]);
    });

    it('passes context to pipeline', async () => {
      const spy = jest.spyOn(beemo, 'startPipeline');

      await beemo.createConfigFiles(MOCK_CONFIG_ARGS, ['foo']);

      expect(spy).toHaveBeenCalledWith(
        expect.objectContaining({
          argv: ['foo', 'bar'],
          drivers: expect.anything(),
        }),
      );
    });

    it('creates for multiple drivers', async () => {
      const spy = jest.spyOn(beemo, 'startPipeline');

      await beemo.createConfigFiles(MOCK_CONFIG_ARGS, ['foo', 'bar', 'baz']);

      expect(spy).toHaveBeenCalled();
      expect((spy.mock.calls[0][0] as any).drivers.size).toBe(3);
    });

    it('creates for all drivers if list is empty', async () => {
      const spy = jest.spyOn(beemo, 'startPipeline');

      await beemo.createConfigFiles(MOCK_CONFIG_ARGS, []);

      expect(spy).toHaveBeenCalledWith(
        expect.objectContaining({
          argv: ['foo', 'bar'],
          drivers: new Set(beemo.tool.getPlugins('driver')),
        }),
      );
      expect((spy.mock.calls[0][0] as any).drivers.size).toBe(3);
    });
  });

  describe('getConfigBlueprint()', () => {
    it('errors if no module', () => {
      expect(() => {
        optimal({}, beemo.getConfigBlueprint());
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if module is defined with env var', () => {
      process.env.BEEMO_CONFIG_MODULE = 'test-boost';

      let opts: any = {};

      expect(() => {
        opts = optimal({}, beemo.getConfigBlueprint());
      }).not.toThrowError();

      expect(opts.module).toBe('test-boost');

      process.env.BEEMO_CONFIG_MODULE = '';
    });
  });

  describe('getConfigModuleRoot()', () => {
    beforeEach(() => {
      beemo.moduleRoot = '';
    });

    it('errors if no module name', () => {
      beemo.tool.config.module = '';

      expect(() => {
        beemo.getConfigModuleRoot();
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if a fake and or missing node module', () => {
      beemo.tool.config.module = 'beemo-this-should-never-exist';

      expect(() => {
        beemo.getConfigModuleRoot();
      }).toThrowErrorMatchingSnapshot();
    });

    it('returns cwd if using @local', () => {
      beemo.tool.config.module = '@local';

      expect(beemo.getConfigModuleRoot()).toBe(process.cwd());
      expect(beemo.moduleRoot).toBe(process.cwd());
      expect(beemo.getConfigModuleRoot()).toBe(beemo.moduleRoot);
    });

    it('returns node module path', () => {
      (fs.existsSync as jest.Mock).mockImplementation(() => true);

      beemo.tool.config.module = 'boost';

      const rootPath = path.join(process.cwd(), 'node_modules/boost');

      expect(beemo.getConfigModuleRoot()).toBe(rootPath);
      expect(beemo.moduleRoot).toBe(rootPath);
      expect(beemo.getConfigModuleRoot()).toBe(beemo.moduleRoot);
    });
  });

  describe('handleCleanupOnFailure()', () => {
    let context: DriverContext;

    beforeEach(() => {
      context = createDriverContext();

      (fs.removeSync as jest.Mock).mockReset();
    });

    it('does nothing if exit code is 0', () => {
      beemo.handleCleanupOnFailure(0, context);

      expect(fs.removeSync).not.toHaveBeenCalled();
    });

    it('does nothing if no config paths', () => {
      beemo.handleCleanupOnFailure(1, context);

      expect(fs.removeSync).not.toHaveBeenCalled();
    });

    it('removes file for each config path', () => {
      context.configPaths = [{ driver: 'foo', path: 'foo' }, { driver: 'bar', path: 'bar' }];

      beemo.handleCleanupOnFailure(1, context);

      expect(fs.removeSync).toHaveBeenCalledWith('foo');
      expect(fs.removeSync).toHaveBeenCalledWith('bar');
    });
  });

  describe('executeDriver()', () => {
    beforeEach(() => {
      // @ts-ignore
      beemo.tool.getPlugin = () => ({
        name: 'foo-bar',
        metadata: { title: 'Foo Bar' },
        getVersion: () => '0.0.0',
      });
    });

    it('triggers `init-driver` event with context', async () => {
      const spy = jest.spyOn(beemo.tool, 'emit');

      await beemo.executeDriver(MOCK_DRIVER_ARGS, 'foo-bar');

      expect(spy).toHaveBeenCalledWith('foo-bar.init-driver', [
        expect.objectContaining({
          argv: ['foo', 'bar'],
          driverName: 'foo-bar',
        }),
        expect.objectContaining({ name: 'foo-bar' }),
      ]);
    });

    it('passes driver name and context to pipeline run', async () => {
      const spy = jest.spyOn(beemo, 'startPipeline');

      await beemo.executeDriver(MOCK_DRIVER_ARGS, 'foo-bar');

      expect(spy).toHaveBeenCalledWith(
        expect.objectContaining({
          argv: ['foo', 'bar'],
          driverName: 'foo-bar',
        }),
      );

      expect(beemo.pipeline!.run).toHaveBeenCalledWith('foo-bar');
    });

    it('passes parallelArgv to context', async () => {
      const spy = jest.spyOn(beemo, 'startPipeline');

      await beemo.executeDriver(MOCK_DRIVER_ARGS, 'foo', [['--foo'], ['bar']]);

      expect(spy).toHaveBeenCalledWith(
        expect.objectContaining({
          parallelArgv: [['--foo'], ['bar']],
        }),
      );
    });

    it('doesnt pipe cleanup routine if `configure.cleanup` is false', async () => {
      beemo.tool.config.configure.cleanup = false;

      await beemo.executeDriver(MOCK_DRIVER_ARGS, 'foo-bar');

      expect(beemo.pipeline!.pipe).toHaveBeenCalledTimes(2);
    });

    it('pipes cleanup routine if `configure.cleanup` is true', async () => {
      beemo.tool.config.configure.cleanup = true;

      await beemo.executeDriver(MOCK_DRIVER_ARGS, 'foo-bar');

      expect(beemo.pipeline!.pipe).toHaveBeenCalledTimes(3);
    });
  });

  describe('executeScript()', () => {
    it('errors if script name is not in kebab case', () => {
      expect(
        beemo.executeScript(MOCK_DRIVER_ARGS, 'Foo_Bar'),
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it('errors if script name starts with a dash', () => {
      expect(beemo.executeScript(MOCK_DRIVER_ARGS, '-foo')).rejects.toThrowErrorMatchingSnapshot();
    });

    it('errors if script name ends with a dash', () => {
      expect(beemo.executeScript(MOCK_DRIVER_ARGS, 'bar-')).rejects.toThrowErrorMatchingSnapshot();
    });

    it('triggers `init-script` event with context', async () => {
      const spy = jest.spyOn(beemo.tool, 'emit');

      await beemo.executeScript(MOCK_DRIVER_ARGS, 'foo-bar');

      expect(spy).toHaveBeenCalledWith('foo-bar.init-script', [
        expect.objectContaining({
          argv: ['foo', 'bar'],
          scriptName: 'foo-bar',
        }),
        'foo-bar',
      ]);
    });
  });

  describe('prepareContext()', () => {
    it('sets extra props', () => {
      expect(beemo.prepareContext(new Context(MOCK_ARGS))).toEqual(
        expect.objectContaining({
          args: MOCK_ARGS,
          argv: ['foo', 'bar'],
          moduleRoot: root,
          cwd: root,
        }),
      );
    });
  });

  describe('startPipeline()', () => {
    beforeEach(() => {
      delete process.beemo;
    });

    it('sets beemo instance on process global', () => {
      expect(process.beemo).toBeUndefined();

      const context = createContext();
      context.args.foo = 123;
      context.args.bar = true;

      beemo.startPipeline(context);

      expect(process.beemo).toEqual({
        context,
        tool: beemo.tool,
      });
    });

    it('registers an exit listener if cleanup is true', async () => {
      beemo.tool.config.configure.cleanup = true;

      beemo.startPipeline(createContext());

      expect(onSpy).toHaveBeenCalledWith('exit', expect.any(Function));
    });

    it('doesnt register exit listener if cleanup is false', async () => {
      beemo.tool.config.configure.cleanup = false;

      beemo.startPipeline(createContext());

      expect(onSpy).not.toHaveBeenCalled();
    });
  });

  describe('scaffold()', () => {
    it('triggers `scaffold` event with context', async () => {
      const spy = jest.spyOn(beemo.tool, 'emit');

      await beemo.scaffold(MOCK_SCAFFOLD_ARGS, 'gen', 'action');

      expect(spy).toHaveBeenCalledWith('beemo.scaffold', [
        expect.objectContaining({
          argv: ['foo', 'bar'],
        }),
        'gen',
        'action',
        '',
      ]);
    });

    it('passes context to pipeline', async () => {
      const spy = jest.spyOn(beemo, 'startPipeline');

      await beemo.scaffold(MOCK_SCAFFOLD_ARGS, 'gen', 'action');

      expect(spy).toHaveBeenCalledWith(
        expect.objectContaining({
          generator: 'gen',
          action: 'action',
        }),
      );
    });
  });
});
