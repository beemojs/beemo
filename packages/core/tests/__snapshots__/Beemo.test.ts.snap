// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Beemo executeScript() errors if script name ends with a dash 1`] = `"Script name must be in kebab case (alphanumeric characters and dashes)."`;

exports[`Beemo executeScript() errors if script name is not in kebab case 1`] = `"Script name must be in kebab case (alphanumeric characters and dashes)."`;

exports[`Beemo executeScript() errors if script name starts with a dash 1`] = `"Script name must be in kebab case (alphanumeric characters and dashes)."`;

exports[`Beemo getConfigModuleRoot() errors if a fake and or missing node module 1`] = `"Module beemo-this-should-never-exist defined in \\"beemo.module\\" could not be found."`;

exports[`Beemo getConfigModuleRoot() errors if no module name 1`] = `"Beemo requires a \\"beemo.module\\" property within your package.json. This property is the name of a module that houses your configuration files."`;

exports[`configBlueprint() errors if no module 1`] = `"Invalid field \\"module\\". Field is required and must be defined."`;
